aptos-move/framework/aptos-framework/sources/AptosGovernance.move

```diff
@@ -0,0 +1,360 @@
+    /// for the CoreFramework (0x1) account.
+    public fun initialize(
+        core_framework: &signer,
+        min_voting_threshold: u128,
+        required_proposer_stake: u128,
+        voting_period_secs: u64,
+    ) {
+        SystemAddresses::assert_core_framework(core_framework);
+
+        Voting::register<GovernanceProposal>(core_framework);
+        move_to(core_framework, GovernanceConfig {
+            voting_period_secs,
+            min_voting_threshold,
+            required_proposer_stake,
+        });
+        move_to(core_framework, GovernanceEvents {
+            create_proposal_events: Event::new_event_handle<CreateProposalEvent>(core_framework),
+            update_config_events: Event::new_event_handle<UpdateConfigEvent>(core_framework),
+            vote_events: Event::new_event_handle<VoteEvent>(core_framework),
+        });
+        move_to(core_framework, VotingRecords {
+            votes: Table::new(),
+        });
+    }
+
+    /// Update the governance configurations. This can only be called as part of resolving a proposal in this same
+    /// AptosGovernance.
+    public fun update_governance_config(
+        _proposal: GovernanceProposal,
+        min_voting_threshold: u128,
+        required_proposer_stake: u128,
+        voting_period_secs: u64,
+    ) acquires GovernanceConfig, GovernanceEvents {
+        let governance_config = borrow_global_mut<GovernanceConfig>(@AptosFramework);
+        governance_config.voting_period_secs = voting_period_secs;
+        governance_config.min_voting_threshold = min_voting_threshold;
+        governance_config.required_proposer_stake = required_proposer_stake;
+
+        let events = borrow_global_mut<GovernanceEvents>(@AptosFramework);
+        Event::emit_event<UpdateConfigEvent>(
+            &mut events.update_config_events,
+            UpdateConfigEvent {
+                min_voting_threshold,
+                required_proposer_stake,
+                voting_period_secs
+            },
+        );
+    }
+
+    /// Create a proposal with the backing `stake_pool`.
+    /// @param execution_hash Required. This is the hash of the resolution script. When the proposal is resolved,
+    /// only the exact script with matching hash can be successfully executed.
+    public(script) fun create_proposal(
+        proposer: &signer,
+        stake_pool: address,
+        execution_hash: vector<u8>,
+    ) acquires GovernanceConfig, GovernanceEvents {
+        let proposer_address = Signer::address_of(proposer);
+        assert!(Stake::is_delegated_voter(stake_pool, proposer_address), Errors::invalid_argument(ENOT_DELEGATED_VOTER));
+
+        // The proposer's stake needs to be at least the required bond amount.
+        let governance_config = borrow_global<GovernanceConfig>(@AptosFramework);
+        // TODO: Remove u128 cast once Coin is migrated to use u128 everywhere.
+        let stake_balance = (Stake::get_active_staked_balance(stake_pool) as u128);
```

Comment:
A general question. Is the Stake here strictly for consensus POS? If so, do we want to give the voting power to more people? Eg: Bob has a lot of coin and haven't delegated for Staking, Bob could also have a right to vote.

Reply:
I mentioned this in the design doc but basically we need a staking mechanism to avoid double voting (balance snapshots are generally prohibitively expensive). This is redundant with validator staking. Users can technically stake their coin via Stake without actually running a node for voting purposes. They just don't receive rewards.

Reply:
is there also pending_active ? do we need that state for voting?

Reply:
Yes this is mentioned in one of the comments around the voting power calculation. For simplicity, we shouldn't take into account pending_active stake and only do active instead. This avoids having to deal with edge cases where the pending_active stake is activated in the next epoch, which might be after the voting period is over.